t/20-function-bqsr_calc.t

```diff
@@ -0,0 +1,159 @@
+    (conf_path          => 't/data/release/config/bqsr_off',
+    runfolder_path      => $runfolder_path,
+    id_run              => 26291,
+    timestamp           => $timestamp);
+  my $ds = $hc->create;
+  is(scalar @{$ds}, 1, 'one definition is returned');
+  isa_ok($ds->[0], 'npg_pipeline::function::definition');
+  is($ds->[0]->excluded, 1, 'function is excluded');
+};
+
+subtest 'bqsr defaulted on' => sub {
+  plan tests => 3;
+
+  my $hc = npg_pipeline::function::bqsr_calc->new
+    (conf_path          => "t/data/release/config/bqsr",
+    runfolder_path      => $runfolder_path,
+    id_run              => 26291,
+    timestamp           => $timestamp,
+    repository          => $dir);
+  my $ds = $hc->create;
+  is(scalar @{$ds}, 12, '12 definitions are returned');
+  isa_ok($ds->[0], 'npg_pipeline::function::definition');
+  is($ds->[0]->excluded, undef, 'function is not excluded');
+};
+
+#######
+# In an array of definitions find and return a definition
+# that corresponds to a position and, optionally, tag index.
+# Use composition property of the definition object.
+#
+sub _find {
+  my ($a, $p, $t) = @_;
+  my $d= first { my $c = $_->composition->get_component(0);
+                 $c->position == $p &&
+                 (defined $t ? $c->tag_index == $t : !defined $c->tag_index) }
+         @{$a};
+  if (!$d) {
+    die "failed to find definition for position $p, tag " . defined $t ? $t : 'none';
+  }
+  return $d;
+}
+
+subtest 'run merge_recompress' => sub {
+  plan tests => 20;
+
+  my $rna_gen;
+  lives_ok {
+  $rna_gen = npg_pipeline::function::bqsr_calc->new(
+    conf_path         => 't/data/release/config/bqsr',
+    run_folder        => $runfolder,
+    runfolder_path    => $runfolder_path,
+    id_run            => 26291,
+    timestamp         => $timestamp,
+    verbose           => 0,
+    repository        => $dir
+  )
+  } 'no error creating an object';
+
+  is ($rna_gen->id_run, 26291, 'id_run inferred correctly');
+
+  my $qc_in  = $dir . q[/novaseq/180709_A00538_0010_BH3FCMDRXX/Data/Intensities/BAM_basecalls_20180805-013153/no_cal/archive/lane4/plex3];
+  my $qc_out = join q[/], $qc_in, q[qc];
+  make_path "$bc_path/archive/tileviz";
+  apply_all_roles($rna_gen, 'npg_pipeline::runfolder_scaffold');
+  $rna_gen->create_product_level();
+
+  my $da = $rna_gen->create();
+
+  ok ($da && @{$da} == 12, sprintf("array of 12 definitions is returned, got %d", scalar@{$da}));
+
+  my $command = qq{$gatk_exec BaseRecalibrator -O $dir/novaseq/180709_A00538_0010_BH3FCMDRXX/Data/Intensities/BAM_basecalls_20180805-013153/no_cal/archive/plex4/26291#4.bqsr_table -I $dir/novaseq/180709_A00538_0010_BH3FCMDRXX/Data/Intensities/BAM_basecalls_20180805-013153/no_cal/archive/plex4/26291#4.cram -R $fasta_dir/GRCh38_15_plus_hs38d1.fa --known-sites $annot_dir/dbsnp_138.hg38.vcf.gz --known-sites $annot_dir/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz};
+
+  my $mem = 2000;
+  my $d = _find($da, 1, 4);
```

Comment:
my $d = $da->[3];
should also work since the definitions follow the sort order of the products (12 merged products ordered by tag index); lines 92-107 can then be removed

Reply:
done